applications/nrf_desktop/src/modules/led_stream.c

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <zephyr.h>
+#include <misc/byteorder.h>
+
+#define MODULE led_stream
+#include "module_state_event.h"
+
+#include <logging/log.h>
+
+#include "config_event.h"
+#include "led_event.h"
+
+#define STREAM_LED_ID 0
+#define INCOMING_LED_COLOR_COUNT 3
+#define STREAM_DATA_SIZE 7
+
+LOG_MODULE_REGISTER(MODULE, CONFIG_DESKTOP_LED_STREAM_LOG_LEVEL);
+
+
+static struct led_effect first_led_stream_effect = {
+	.step_count = 1,
+};
+static struct led_effect second_led_stream_effect = {
+	.step_count = 1,
+};
+static struct led_effect_step steps_queue[CONFIG_DESKTOP_LED_STREAM_QUEUE_SIZE];
+static u8_t rx_idx;
+static u8_t tx_idx;
+static bool first_sent;
+static bool initialized;
+static bool first;
+
+
+static size_t next_index(size_t index)
+{
+	return (index + 1) % CONFIG_DESKTOP_LED_STREAM_QUEUE_SIZE;
+}
+
+static bool queue_data(const struct event_dyndata *dyndata)
+{
+	if (dyndata->size != STREAM_DATA_SIZE) {
+		LOG_WRN("Invalid stream data size (%"
+			PRIu8 ")", dyndata->size);
+		return false;
+	}
+
+	LOG_INF("Insert data to effect");
+
+	size_t pos = 0;
+
+	memcpy(steps_queue[rx_idx].color.c, &dyndata->data[pos],
+	       CONFIG_DESKTOP_LED_COLOR_COUNT);
+	pos += INCOMING_LED_COLOR_COUNT;
+	steps_queue[rx_idx].substep_count = sys_get_le16(&dyndata->data[pos]);
+	pos += sizeof(steps_queue[rx_idx].substep_count);
+	steps_queue[rx_idx].substep_time = sys_get_le16(&dyndata->data[pos]);
+
+	rx_idx = next_index(rx_idx);
+
+	return true;
+}
+
+static void send_effect(struct led_effect *effect, size_t led_id)
+{
+	LOG_INF("Send_effect to leds");
+
+	struct led_event *led_event = new_led_event();
+
+	led_event->stream = true;
+	led_event->led_id = led_id;
+	led_event->led_effect = effect;
+
+	__ASSERT_NO_MSG(led_event->led_effect->steps);
+
+	EVENT_SUBMIT(led_event);
+}
+
+static size_t count_free_places(void)
+{
+	size_t len = (CONFIG_DESKTOP_LED_STREAM_QUEUE_SIZE + rx_idx - tx_idx)
+		     % CONFIG_DESKTOP_LED_STREAM_QUEUE_SIZE;
+	return CONFIG_DESKTOP_LED_STREAM_QUEUE_SIZE - len;
+}
+
+static bool store_data(const struct event_dyndata *data)
+{
+	size_t free_places = count_free_places();
+
+	if (free_places > 0) {
+		LOG_INF("Insert data on position %" PRIu8
+			", free places %zu", rx_idx, free_places);
+
+		if (!queue_data(data)) {
+			return false;
+		}
+	} else {
+		LOG_WRN("Queue is full - dropping incoming step");
+	}
+	return true;
+}
+
+static void send_data_from_queue(void)
+{
+	size_t free_places = count_free_places();
+
+	if (free_places < CONFIG_DESKTOP_LED_STREAM_QUEUE_SIZE) {
+		LOG_INF("Sending data");
+
+		if (first) {
```

Comment:
this is what I meant when I commented two led effect variables
see the following
```
led_stream_effect[cur_effect_idx].steps = &steps[x];
send_effect(&led_stream_effect[cur_effect_idx], ID);
cur_effect_idx = 1 - cur_effect_idx;
```

Reply:
done